### PR TITLE
[Enhance]: Speedup yolov3 inference.

### PR DESCRIPTION
**Pull request by RangiLyu**
## Motivation

Speed up yolov3 inference by refactoring the `get_bboxes` logic.

## Modification

To speed up the inference time, I refactored the following part:
1. Move onnx export logic to `onnx_export` function. Simplify the `get_bboxes` logic. Speed up yolov3 mbv2 320 about 3 FPS.
2. Improve code parallelism. Replace the for loop of multi-level feature map decode with a flatten tensor decode. Speed up yolov3 mbv2 320 about 6 FPS.
3. Simplify the bbox decode in yolo bbox coder. Speed up yolov3 mbv2 320 about 4 FPS.

Finally improved about 10FPS.

* *Note: Tested on a single 1060 6G GPU with torch1.8 cuda11.1*

## Results

|           | yolov3 d53 608 | yolov3 d53 416 | yolov3 d53 320 | yolov3 mbv2 416 | yolov3 mbv2 320 |
|---------|----------------|----------------|----------------|-----------------|-----------------|
| Before  | 23.2fps        | 44.9fps        | 53.2fps        | 79.7fps         | 93.9fps         |
| After     | 24.2fps        | 48.5fps        | 57.4fps        | 91.5fps         | 106.2fps        |

## BC-breaking (Optional)

None

### CODE REVIEW

File: mmdet/models/dense_heads/yolo_head.py [hunk 1]
```diff
@@ -215,200 +215,68 @@ def get_bboxes(self,
...
+
+        if with_nms and (flatten_objectness.size(0) == 0):
             return torch.zeros((0, 5)), torch.zeros((0, ))
 
         if rescale:
-            batch_mlvl_bboxes /= batch_mlvl_bboxes.new_tensor(
+            flatten_bboxes /= flatten_bboxes.new_tensor(
                 scale_factors).unsqueeze(1)
 
         # In mmdet 2.x, the class_id for background is num_classes.
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/yolo_head.py:254]: This comment seems to be unnecessary, can be removed.

File: mmdet/models/dense_heads/yolo_head.py [hunk 2]
```diff
@@ -215,200 +215,68 @@ def get_bboxes(self,
...
-
-        if with_nms and (batch_mlvl_conf_scores.size(0) == 0):
+        num_imgs = len(img_metas)
+        featmap_sizes = [pred_map.shape[-2:] for pred_map in pred_maps]
+
+        mlvl_anchors = self.anchor_generator.grid_anchors(
+            featmap_sizes, pred_maps[0].device)
+        flatten_preds = []
+        flatten_strides = []
+        for pred, stride in zip(pred_maps, self.featmap_strides):
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/yolo_head.py:229]: Suggest leaving a comment in the code to indicate which part really affects the speed to remind others to take care of that part.

File: mmdet/core/bbox/coder/yolo_bbox_coder.py [hunk 3]
```diff
@@ -70,21 +70,13 @@ def decode(self, bboxes, pred_bboxes, stride):
...
-        y_center = (bboxes[..., 1] + bboxes[..., 3]) * 0.5
-        w = bboxes[..., 2] - bboxes[..., 0]
-        h = bboxes[..., 3] - bboxes[..., 1]
-        # Get outputs x, y
-        x_center_pred = (pred_bboxes[..., 0] - 0.5) * stride + x_center
-        y_center_pred = (pred_bboxes[..., 1] - 0.5) * stride + y_center
-        w_pred = torch.exp(pred_bboxes[..., 2]) * w
-        h_pred = torch.exp(pred_bboxes[..., 3]) * h
-
+        xys = (bboxes[..., :2] +
```
Comment (hhaAndroid) [inline mmdet/core/bbox/coder/yolo_bbox_coder.py:74]: xy_center

File: mmdet/models/dense_heads/yolo_head.py [hunk 4]
```diff
@@ -215,200 +215,68 @@ def get_bboxes(self,
...
+            pred = pred.permute(0, 2, 3, 1).reshape(num_imgs, -1,
+                                                    self.num_attrib)
+            pred[..., :2].sigmoid_()
+            flatten_preds.append(pred)
+            flatten_strides.append(
+                pred.new_tensor(stride).expand(pred.size(1)))
+
+        flatten_preds = torch.cat(flatten_preds, dim=1)
+        flatten_bbox_preds = flatten_preds[..., :4]
+        flatten_objectness = flatten_preds[..., 4].sigmoid()
```
Comment (hhaAndroid) [inline mmdet/models/dense_heads/yolo_head.py:239]: Why can't the sigmoid be unified in the front but calculated separately?

File: mmdet/models/dense_heads/yolo_head.py [hunk 5]
```diff
@@ -215,200 +215,68 @@ def get_bboxes(self,
...
-                tuple(mlvl_bs)
-                for mlvl_bs in zip(batch_mlvl_bboxes, batch_mlvl_scores,
-                                   batch_mlvl_conf_scores)
-            ]
+        padding = flatten_bboxes.new_zeros(num_imgs, flatten_bboxes.shape[1],
+                                           1)
+        flatten_cls_scores = torch.cat([flatten_cls_scores, padding], dim=-1)
+
+        det_results = []
+        for (bboxes, scores, objectness) in zip(flatten_bboxes,
```
Comment (hhaAndroid) [inline mmdet/models/dense_heads/yolo_head.py:261]: If this process is referenced from other repo, then citations and instructions need to be added.

Comment (hhaAndroid) [reply mmdet/models/dense_heads/yolo_head.py:261]: It seems that the order has been changed, and the overall logic has not changed?

File: mmdet/models/dense_heads/yolo_head.py (hunk 2 above)
Comment (hhaAndroid) [reply mmdet/models/dense_heads/yolo_head.py:229]: Very reasonable. The most important thing is why the original process is slow？

File: mmdet/models/dense_heads/yolo_head.py (hunk 4 above)
Comment (RangiLyu) [reply mmdet/models/dense_heads/yolo_head.py:239]: It seems that using inplace sigmoid is a little bit slower than create a new tensor. Maybe because of the memory continuity.

**Review by ZwwWayne — APPROVED** [review]
